Framework/DataHandling/src/MaskDetectors.cpp

```diff
@@ -269,83 +399,143 @@ void MaskDetectors::execPeaks(PeaksWorkspace_sptr WS) {
     }
   }
 }
-/**
- * Convert a list of spectra numbers into the corresponding workspace indices
+
+/* Convert a list of spectra numbers into the corresponding workspace indices.
+ *
  * @param indexList :: An output index list from the given spectra list
  * @param spectraList :: A list of spectra numbers
  * @param WS :: The input workspace to be masked
+ * @param range_info :: tuple containing the range of spectra to process and
+ *                      Boolean indicating if these ranges are defined
  */
 void MaskDetectors::fillIndexListFromSpectra(
     std::vector<size_t> &indexList, const std::vector<specnum_t> &spectraList,
-    const API::MatrixWorkspace_sptr WS) {
-  // Convert the vector of properties into a set for easy searching
-  std::set<int64_t> spectraSet(spectraList.begin(), spectraList.end());
-  // Next line means that anything in Clear the index list first
-  indexList.clear();
-  indexList.reserve(WS->getNumberHistograms());
-
-  for (int i = 0; i < static_cast<int>(WS->getNumberHistograms()); ++i) {
-    const specnum_t currentSpec = WS->getSpectrum(i).getSpectrumNo();
-    if (spectraSet.find(currentSpec) != spectraSet.end()) {
-      indexList.push_back(i);
+    const API::MatrixWorkspace_sptr WS,
+    const std::tuple<size_t, size_t, bool> &range_info) {
+
+  std::vector<size_t> tmp_index;
+  size_t startIndex = std::get<0>(range_info);
+  size_t endIndex = std::get<1>(range_info);
+  bool range_constrained = std::get<2>(range_info);
+
+  if (range_constrained) {
+    constrainIndexInRange(indexList, tmp_index, startIndex, endIndex);
+  } else {
+    tmp_index.swap(indexList);
```

Comment:
swap modifies indexList and return the result. Assign will then work without tmp vector but needs internal reallocations.